druid/src/env.rs

```diff
@@ -133,6 +133,35 @@ pub enum KeyOrValue<T> {
     Key(Key<T>),
 }
 
+/// A trait for anything that can resolve a value of some type from the [`Env`].
+///
+/// This is a generalization of the idea of [`KeyOrValue`], mostly motivated
+/// by wanting to improve the API used for checking if items in the [`Env`] have changed.
+///
+/// [`Env`]: struct.Env.html
+/// [`KeyOrValue`]: enum.KeyOrValue.html
+pub trait KeyLike<T> {
+    /// Returns `true` if this item has changed between the old and new [`Env`].
+    ///
+    /// [`Env`]: struct.Env.html
+    fn changed(&self, old: &Env, new: &Env) -> bool;
+}
+
+impl<T: ValueType> KeyLike<T> for Key<T> {
+    fn changed(&self, old: &Env, new: &Env) -> bool {
+        !old.get_untyped(self).same(new.get_untyped(self))
+    }
+}
+
+impl<T> KeyLike<T> for KeyOrValue<T> {
+    fn changed(&self, old: &Env, new: &Env) -> bool {
+        match self {
+            KeyOrValue::Concrete(_) => false,
+            KeyOrValue::Key(key) => !old.get_untyped(key).same(new.get_untyped(key)),
```

Comment:
Same as above.

Reply:
huh, is your panic that some key is missing from the env?

Reply:
Yeah, pretty sure that's it.

I am trying to change the color of a menu label based on the current selection and hot state.
I might be doing something wrong, because I'm in the middle of a upgrade to the latest master (including this PR) and haven't yet looked at all the changes and only tried to get everything in a working state again.

I've created a somewhat "minimal" example repo for this, if you want to take a look:
https://github.com/mmitteregger/druid-example/blob/master/src/widget/menu/menu_item.rs#L123-L125

Reply:
In your example, I wouldn't use a `Key` at all for setting the color; I would just explicitly pass the color you want to the label.

I also wouldn't use `WidgetPod` for your label; if you are just using it for layout, you can draw the label where you want in `Paint` by using `Label::draw_at` (this is new).

Reply:
Having dug into this a bit more: you have three keys in the `Env`, `MENU_ITEM_SELECTED_COLOR` and three other variants. you then also have a key `Color` that you're setting on the label, and then you're dynamically updating this key based on the other keys.

when you set a `Key` as the value for something like text color, it means that when drawing we will figure out what that key's current color is, and then we'll use that color. So if you want to change the color, one option is to never change the key assigned to the label, and instead change the value for that key in the `Env`. An alternative is to change the key assigned to the label; for instance when your text is selected, you could do `label.set_text_color(MENU_ITEM_SELECTED_COLOR)`. In your particular case, I think this is probably the easiest approach?

The rationale for this is that you could have different themes that provide different values for various keys, and it would 'just work'. Similarly, if you were using built-in keys we could support things like dark mode, or high-contrast colors for accessibility, and it would 'just work'.

Your crash is happening because the `COLOR` key you're using doesn't always exist in the `Env`;https://linebender.org/druid/env.html#keys-values-and-themes. I think this crash is okay; by the time `update` happens, any key you've assigned to a label must exist, and we crashing if it doesn't is reasonable behaviour.